### PR TITLE
Exile Implant Revamp: Borders are Closed

### PR DESCRIPTION
**Pull request by GreatBigFailure**
tl;dr: If you implant someone it won't do anything until they first leave the Z they were implanted on. Then they can't return to it.

I’ve reworked exile implants to be more useful, more flexible, and I think more fun. 

**Why**
The old exile implants just bound you to specifically Z5. This meant you could only explore the roid and on roid station they were flat useless. They also didn’t actually un-exile you when you had the implant removed. This is a good idea but poor execution for a non-lethal way to handle traitors that isn’t perma. We need more of those so here’s another. 

**What**

Exile implants now record the Z you were implanted on. When implanting someone you can also add a disable phrase. The implant doesn’t do anything until they’ve first left the Z level. Once their Z has changed they’ll receive a message whenever re-entering the banned Z. If they don’t leave within 10 seconds they’ll be randomly teleported to a tile within any Z besides centcomm and their banned Z.

The disable phrase or EMPs jaunts the implant back to the tile you were first implanted. So it can work as a notifier that it was removed. EMPs aren’t guaranteed however and can cause the crude wormhole tech in the implant to malfunction.

This is a basic version since it’s already a fairly big addition and could impact players as is. I've added exile implants to cargo as well. I have plans for different versions but for now let's try this out.

:cl:
 * rscadd: Exile implants now work properly as a non-lethal prisoner removal method.

### CODE REVIEW

**Comment by SonixApache** [comment]
Be *sure* to test this with regards to nuke disk fuckery.

**Comment by Dacendeth** [comment]
wont the random teleport sometimes teleport them inside the roid and they'd just be stuck?

**Comment by GreatBigFailure** [comment]
> Be _sure_ to test this with regards to nuke disk fuckery.

I knew I forgot to test something. I'll do that in a few minutes.

**Comment by cloutcode** [comment]
>not delivering a lethal dose of neurotoxin to whoever dares to try border hopping for their sins

**Comment by GreatBigFailure** [comment]
> wont the random teleport sometimes teleport them inside the roid and they'd just be stuck?

I don't know how that never happened while I was testing but I 100% do not have any safeties in place for that. I'll add some, shouldn't be hard.

**Comment by DrSnips** [comment]
I don't think any changes to gateway is needed. With antags they are if anything showing you mercy by letting you drift in space rather than just killing you. 200 is also a little steep for the basic level of implants if you ask me.

> Exile implants now record the Z you were implanted on

Does this mean Z is the one they are locked TO or locked FROM?

**Comment by GreatBigFailure** [comment]
My mindset with the gateway is that since this is a revamp and not some new exciting item people likely aren't going to think to seek it out. Seeing it on the cargo list might change that. Exile implants also aren't even associated with the gateway anymore so it's just weird that they're in there. 
200 was a spitball number but you're right, a bit lower at least. Much cheaper if people want me to leave it how it is, a fair bit more expensive if they want the bluespace crystal thing or similar. 

> Does this mean Z is the one they are locked TO or locked FROM?

Locked FROM. 
Once you leave the Z you were implanted on the implant won't let you return to it. So you could have a Trader ban a crook from their base or have a Xenoarch ban a miner who destroyed all their digs from the roid. I doubt either of those will ever happen but it's now flexible is the point.

**Comment by GreatBigFailure** [comment]
There, can't be teleported into a wall anymore. 

I also tested the disk and it's fine. It just prevents you from teleporting. Any other interaction with the disk would either require it to not be on station Z or just be a normal man trying to carry it off the Z and failing. Tried both teleporting out of Z1 by being banned and teleporting to other Z levels with EMP malfunction.

**Comment by Viciousarus** [comment]
>Teleporting them
>Not gibbing them
Weak implant

**Comment by Dacendeth** [comment]
Wont someone be able to return if they just wear a tinfoil hat, as it prevents teleporting?

**Comment by GreatBigFailure** [comment]
> Wont someone be able to return if they just wear a tinfoil hat, as it prevents teleporting?

Yep, but they'll get stunned every 10 seconds and start receiving minor damage slowly. Each teledeport has a 50 50 chance of causing 5 clone and 5 brute. I'm thinking of removing that though.

**Comment by TGFH** [comment]
Why you would remove the implants from the gateway if the goal is to get people to actually use the implants? I get that the gateway thing is defunct but walking to the gateway for free implants > going to cargo and paying + waiting for the cargonians to send the crate to the wrong department.

**Comment by GreatBigFailure** [comment]
Mostly it just bugs me that they're there, I admit. I can ask a mapper to just put them in sec instead then add them to cargo as well? Basically how loyalty implants currently work. Gateway is such a goofy place for them.

**Comment by tuypo1** [comment]
> not delivering a lethal dose of neurotoxin to whoever dares to try border hopping for their sins

Problem with that is you would be floating through space and if you happened to hit the station z you would just die

**Comment by GreatBigFailure** [comment]
>not delivering a lethal dose of neurotoxin to whoever dares to try border hopping for their sins

>Weak implant

The issue, like Tuypo said is that you don't control what Z level you enter when transitioning. There's a bunch of different factors that might see them accidentally return to Z1. I also wanted to make it flexible since that's what makes the best stories. This opens possibilities for dumb stuff like a teleport wizard that exiles himself then teleports in, bullshits around for 10 seconds, then gets auto-teleported away. 
Besides, it's meant to be non-lethal.

**Comment by SonixApache** [comment]
would be a lot better if people just couldn't drift onto their banned z-level in the first place

**Comment by GreatBigFailure** [comment]
> would be a lot better if people just couldn't drift onto their banned z-level in the first place

Why do you say that? 
The teleport is for flexibility and **_emergent gameplay_**. It's also to make it hard to get rid of but not necessarily a 100% permanent timeout chair from video games. I'm open to being convinced though.

**Comment by SonixApache** [comment]
- people will get 'punished' for something they cannot really prevent - drifting onto the main Z
- people will actively try to launch themselves onto the banned z-level to teleport, hoping they'll hit something nice like a vault
 - people can and will get stuck in extremely gay places such as roid tunnels after being teleported away
 - can't shoot people out the mass driver because they'll teleport away before you can even hit the button
- can't banish people through the gateway because they can just walk out the door and get teleported to the roid or something
 - since there's no z-limit apart from centcomm, people could straight up be teleported to anywhere in an away mission

I'm cool with the teleport functionality, it applies well to when people hop onto shuttles or find some other fucked up way to get back onto their banned Z-level.
But really, they should just be unable to get there in the first place in as many ways as possible. Subconscious evasion, or something.

**Comment by GreatBigFailure** [comment]
The teleport isn't much of a punishment, that's one reason I'm thinking I should get rid of the clone damage. 
That is a good point about actively trying to get to the banned Z for intentional warping. Originally that's why I added the clone damage but now I'm torn. 
The tunnels are a problem I hadn't considered. Fair point.
The mass driver and gateway are incorrect though. They don't teleport away until they've first left the banned Z once. They're fair game aside from that. 
>Teleported to anywhere in an away mission

Wait what. I thought away missions were vaults spawned on z2?

I'll look into how I would go about what you've suggested. I can see a lot of reason for it but it also reduces the flexibility a bit so I'm on the fence.

**Comment by SonixApache** [comment]
Away missions are spawned on Z-6.
Good point on the mass driver, but my gateway issue still stands; you can throw them through the hoop, but they can then just walk back through the hoop on their end and end up somewhere else.

**Comment by GreatBigFailure** [comment]
>Away missions are spawned on Z-6.

I might need to just make it only teleport them to space turfs then. I guess I could do an area check but I've also realized they have a very, very small chance of just getting trapped forever in a vault secret room or something. That would also fix the mining tunnel issue. 
Away missions might have teleport prevention? I assume they do but I haven't checked. Either way I might add the space turf thing because it would fix so many potential issues, including most you've presented, in one clean move.

**Comment by GreatBigFailure** [comment]
Exile implants will now only teleport you to space tiles, they can still teleport you right next to a vault but not into one, an away mission, or a mining tunnel. 

I also added exile lockboxes to cargo. They're 150 credits but a lockbox has 3 implants in it. Gateway implants were replaced by 2 lockboxes, same amount of implants but it's at least less weird now. Lockbox access is armory.

Finally I changed the brute/clone damage from a 50% chance to cause 5 of each to a 100% chance to cause up to 5 of each, randomly.

File: code/game/objects/items/weapons/implants/exile2.dm [hunk 1]
```diff
@@ -1,39 +1,125 @@
-/**
-* Exile Implants 2.0
-*
-* Since away missions are fucked, here's an alternative implementation.
-*
-* Instead of confining someone within an away mission, this locks someone to the asteroid.
-*/
+
+//There now it's fun
```
Comment (DamianX) [inline code/game/objects/items/weapons/implants/exile2.dm]:
```suggestion
```

File: code/game/objects/items/weapons/implants/exile2.dm [hunk 2]
```diff
@@ -1,39 +1,125 @@
...
+	src.imp = new /obj/item/weapon/implant/exile(src)
 	..()
 	update()
 	return
 
-
 /obj/item/weapon/implant/exile
 	name = "exile"
-	desc = "Prevents you from returning from the asteroid"
+	desc = "Prevents returning to where you were implanted"
```
Comment (DamianX) [inline code/game/objects/items/weapons/implants/exile2.dm]:
```suggestion
	desc = "Prevents returning to where you were implanted."
```

File: code/game/objects/items/weapons/implants/exile2.dm [hunk 3]
```diff
@@ -1,39 +1,125 @@
...
+	var/illegalZ = null
+	var/mob/living/theExile = null
```
Comment (DamianX) [inline code/game/objects/items/weapons/implants/exile2.dm:15]: implants already have a var to track the mob they're implanted into, called `imp_in`, which is a terrible name but it's probably better than using a special snowflake var for the same purpose.

File: code/game/objects/items/weapons/implants/exile2.dm [hunk 4]
```diff
@@ -1,39 +1,125 @@
...
 
-/obj/item/weapon/implant/exile/implanted(mob/M)
-	if(!istype(M, /mob/living/carbon))
-		return 0
-	var/mob/living/carbon/I = M
-	to_chat(I, "<span class='notice'>You shiver as you feel a weak bluespace void surround you.</span>")
-	I.locked_to_z = ASTEROID_Z
+/obj/item/weapon/implant/exile/implanted(mob/source as mob)
+	theExile = source
+	disablePhrase = input("Choose phrase to disable:") as text
```
Comment (DamianX) [inline code/game/objects/items/weapons/implants/exile2.dm]: `input`s without a cancel option are annoying.
calling `input` without passing a mob will default to usr, which is bad.
use this instead:
```suggestion
	disablePhrase = stripped_input(user, "Choose phrase to disable:")
```
The phrasing is a bit awkward as well.
And you should now handle the case in which disablePhrase ends up being null, and `return 0` accordingly

File: code/game/objects/items/weapons/implants/exile2.dm [hunk 5]
```diff
@@ -1,39 +1,125 @@
...
+<b>For non-permanent use a disable phrase may be assigned on application.<BR>"}
 	return dat
 
-/obj/item/weapon/implant/exile/implanted(mob/M)
-	if(!istype(M, /mob/living/carbon))
-		return 0
-	var/mob/living/carbon/I = M
-	to_chat(I, "<span class='notice'>You shiver as you feel a weak bluespace void surround you.</span>")
-	I.locked_to_z = ASTEROID_Z
+/obj/item/weapon/implant/exile/implanted(mob/source as mob)
```
Comment (DamianX) [inline code/game/objects/items/weapons/implants/exile2.dm]: It looks like the proc definition `/obj/item/weapon/implant/proc/implanted(var/mob/source)` only supports a single parameter but in truth callers of this proc pass a second argument. See implanter.dm line 33. All definitions and calls of this proc should be adjusted to reflect this, like so:
```suggestion
/obj/item/weapon/implant/exile/implanted(mob/source, mob/user)
```

File: code/game/objects/items/weapons/implants/exile2.dm [hunk 6]
```diff
@@ -1,39 +1,125 @@
...
+			theExile.adjustBruteLoss(rand(0,5))
+			theExile.adjustCloneLoss(rand(0,5)) //Uh oh it missed a few chromosomes
+		else
+			to_chat(theExile, "<span class='notice'>Your [part] emits a feint chime. The tingling stops.</span>")
+		beingDeported = FALSE
+
+/obj/item/weapon/implant/exile/emp_act()
+	if(malfunction)
+		return
+	malfunction = 1
```
Comment (DamianX) [inline code/game/objects/items/weapons/implants/exile2.dm:78]: pick one:
- consistently set this to fixed values like here
- consistently decrease/increase this value, like you did in the spawn()

File: code/game/objects/items/weapons/implants/exile2.dm [hunk 7]
```diff
@@ -1,39 +1,125 @@
...
+	var/mob/living/M = imp_in
+	#define FREEDOM 1
+	#define RANDOM_TELEPORT 2
+	#define IMPLANTED_SITE_PORT 3
+	switch(pick(1,2,3))
```
Comment (DamianX) [inline code/game/objects/items/weapons/implants/exile2.dm]:
```suggestion
	switch(pick(FREEDOM,RANDOM_TELEPORT,IMPLANTED_SITE_PORT))
```

File: code/game/objects/items/weapons/implants/exile2.dm [hunk 8]
```diff
@@ -1,39 +1,125 @@
...
+		else
+			to_chat(theExile, "<span class='notice'>Your [part] emits a feint chime. The tingling stops.</span>")
+		beingDeported = FALSE
+
+/obj/item/weapon/implant/exile/emp_act()
+	if(malfunction)
+		return
+	malfunction = 1
+	var/mob/living/M = imp_in
+	#define FREEDOM 1
```
Comment (DamianX) [inline code/game/objects/items/weapons/implants/exile2.dm:79]: `#undef` these once you don't need them anymore

File: code/game/objects/items/weapons/implants/exile2.dm [hunk 9]
```diff
@@ -1,39 +1,125 @@
...
+	illegalZ = theExile.z
+	siteOfImplant = get_turf(theExile)
+	zlevels -= illegalZ
+	to_chat(theExile, "<span class='notice'>You shiver as you feel a weak, unsettling film surround you.</span>")
+	zBanishment_key = theExile.on_moved.Add(src, "zBan")
 	return 1
 
+/obj/item/weapon/implant/exile/proc/zBan()
+	var/turf/T = get_turf(src)
+	switch(beenSpaced)
```
Comment (DamianX) [inline code/game/objects/items/weapons/implants/exile2.dm]: this switch is awkward, just use `if()` and `else()`

File: code/game/objects/items/weapons/implants/exile2.dm [hunk 10]
```diff
@@ -1,39 +1,125 @@
...
+				beingDeported = TRUE
+				teleDeport()
+
+/obj/item/weapon/implant/exile/proc/teleDeport()
+	to_chat(theExile, "<span class='notice'>Your insides churn and your skin tingles. Your [part] is emitting a low hum.</span>")
+	spawn(10 SECONDS)
+		var/turf/T = get_turf(src)
+		if(T.z == illegalZ)
+			var/warpZ = pick(zlevels)
+			var/warpTo = locate(rand(5,world.maxx - 10), rand(5, world.maxy - 10), warpZ)
```
Comment (DamianX) [inline code/game/objects/items/weapons/implants/exile2.dm]: `TRANSITIONEDGE` is how far away from the map border one teleports to a different level, you could use it instead of magic numbers

**Review by DamianX — CHANGES_REQUESTED** [review]

**Comment by DamianX** [comment]
> The disable phrase or EMPs jaunts the implant back to the tile you were first implanted.

I don't understand why you would do this.
It doesn't make sense to me that the implant would teleport back to the original location.
I think upon hearing the specified phrase the implant should just stop functioning, and/or be removed.

**Comment by GreatBigFailure** [comment]
>It doesn't make sense to me that the implant would teleport back to the original location.

It's so that, if you put the extra effort in to implant them somewhere you go a lot, you can see it there and know they've disabled it. It's so that exiling an antag doesn't then give them the element of surprise. It makes them more tempting/easy to use. 
It also helps explain the EMP effect of jaunting them back to their implant location. I wanted that because it will give them 10 seconds of panic/opportunity and just be funny.

**Comment by SonixApache** [comment]
you gotta remove the file from the rsc if it's no longer being loaded
open up DM, should do it

File: code/game/objects/items/weapons/implants/exile2.dm (hunk 3 above)
Comment (GreatBigFailure) [reply code/game/objects/items/weapons/implants/exile2.dm:15]: I can't change that, the implant requires typecasting to work. Won't even register as implanted in the person if I use imp_in. Even the previous version of the implant had a typecast variable to replace imp_in.
